File: tensorflow/python/data/experimental/kernel_tests/optimization/inject_prefetch_test.py
# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Tests for the `AutotuneBuffers` rewrite."""
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

from absl.testing import parameterized

from tensorflow.python.compat import compat
from tensorflow.python.data.experimental.ops import testing
from tensorflow.python.data.kernel_tests import test_base
from tensorflow.python.data.ops import dataset_ops
from tensorflow.python.framework import combinations
from tensorflow.python.platform import test


class InjectPrefetchTest(test_base.DatasetTestBase, parameterized.TestCase):

  def _enable_autotune_buffers(self, dataset):
    options = dataset_ops.Options()
    options.experimental_optimization.autotune_buffers = True
    return dataset.with_options(options)

  @combinations.generate(test_base.default_test_combinations())
  def testParallelMap(self):
    dataset = dataset_ops.Dataset.range(100)
    parallel_map = "ParallelMap"
    if compat.forward_compatible(2020, 3, 6):
      parallel_map = "ParallelMapV2"
    dataset = dataset.apply(
        testing.assert_next([parallel_map, "Prefetch", "FiniteTake"]))
    dataset = dataset.map(
        lambda x: x + 1, num_parallel_calls=dataset_ops.AUTOTUNE)
    dataset = dataset.take(50)
    dataset = self._enable_autotune_buffers(dataset)
    self.assertDatasetProduces(dataset, range(1, 51))

  @combinations.generate(test_base.default_test_combinations())
  def testMapAndBatch(self):
    dataset = dataset_ops.Dataset.range(100)
    dataset = dataset.apply(
        testing.assert_next(["MapAndBatch", "Prefetch", "FiniteTake"]))
    dataset = dataset.map(
        lambda x: x + 1, num_parallel_calls=dataset_ops.AUTOTUNE)
    dataset = dataset.batch(10)
    dataset = dataset.take(5)
    dataset = self._enable_autotune_buffers(dataset)
    self.assertDatasetProduces(
        dataset, [list(range(i + 1, i + 11)) for i in range(0, 50, 10)])

  @combinations.generate(test_base.default_test_combinations())
  def testParallelInterleave(self):
    dataset = dataset_ops.Dataset.range(100)
    parallel_interleave = "ParallelInterleaveV3"
    if compat.forward_compatible(2020, 3, 6):
      parallel_interleave = "ParallelInterleaveV4"
    dataset = dataset.apply(
        testing.assert_next([parallel_interleave, "Prefetch", "FiniteTake"]))
    dataset = dataset.interleave(
        lambda x: dataset_ops.Dataset.from_tensors(x + 1),
        num_parallel_calls=dataset_ops.AUTOTUNE)
    dataset = dataset.take(50)
    dataset = self._enable_autotune_buffers(dataset)
    self.assertDatasetProduces(dataset, range(1, 51))

  @combinations.generate(test_base.default_test_combinations())
  def testChainedParallelDatasets(self):
    dataset = dataset_ops.Dataset.range(100)
    parallel_interleave = "ParallelInterleaveV3"
    if compat.forward_compatible(2020, 3, 6):
      parallel_interleave = "ParallelInterleaveV4"
    parallel_map = "ParallelMap"
    if compat.forward_compatible(2020, 3, 6):
      parallel_map = "ParallelMapV2"
    dataset = dataset.apply(
        testing.assert_next([
            parallel_map, "Prefetch", parallel_interleave, "Prefetch",
            "MapAndBatch", "Prefetch", "FiniteTake"
        ]))
    dataset = dataset.map(
        lambda x: x + 1, num_parallel_calls=dataset_ops.AUTOTUNE)
    dataset = dataset.interleave(
        lambda x: dataset_ops.Dataset.from_tensors(x + 1),
        num_parallel_calls=dataset_ops.AUTOTUNE)
    dataset = dataset.map(
        lambda x: x + 1, num_parallel_calls=dataset_ops.AUTOTUNE)
    dataset = dataset.batch(1)
    dataset = dataset.take(50)
    dataset = self._enable_autotune_buffers(dataset)
    self.assertDatasetProduces(dataset, [[i] for i in range(3, 53)])

  @combinations.generate(test_base.default_test_combinations())
  def testNoRegularMap(self):
    dataset = dataset_ops.Dataset.range(100)
    dataset = dataset.apply(testing.assert_next(["Map", "FiniteTake"]))
    dataset = dataset.map(lambda x: x + 1).take(50)
    dataset = self._enable_autotune_buffers(dataset)
    self.assertDatasetProduces(dataset, range(1, 51))


if __name__ == "__main__":
  test.main()

File: tensorflow/core/grappler/optimizers/data/inject_prefetch_test.cc
/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow/core/grappler/optimizers/data/inject_prefetch.h"

#include "tensorflow/core/framework/attr_value_util.h"
#include "tensorflow/core/framework/function_testlib.h"
#include "tensorflow/core/framework/tensor_testutil.h"
#include "tensorflow/core/grappler/grappler_item.h"
#include "tensorflow/core/grappler/optimizers/data/graph_test_utils.h"
#include "tensorflow/core/grappler/optimizers/data/graph_utils.h"
#include "tensorflow/core/lib/core/status_test_util.h"
#include "tensorflow/core/platform/test.h"

namespace tensorflow {
namespace grappler {
namespace {

using test::function::NDef;

constexpr char kOptionsDataset[] = "OptionsDataset";
constexpr char kParallelMapDataset[] = "ParallelMapDatasetV2";
constexpr char kPrefetchDataset[] = "PrefetchDataset";

Status Optimize(InjectPrefetch &optimizer, const GrapplerItem &item,
                GraphDef *output, bool autotune) {
  RewriterConfig_CustomGraphOptimizer config;
  if (autotune) {
    (*config.mutable_parameter_map())["autotune"].set_s("true");
  } else {
    (*config.mutable_parameter_map())["autotune"].set_s("false");
  }
  TF_RETURN_IF_ERROR(optimizer.Init(&config));
  return optimizer.Optimize(nullptr, item, output);
}

Status OptimizeWithInjectPrefetch(const GrapplerItem &item, GraphDef *output,
                                  bool autotune) {
  InjectPrefetch optimizer;
  return Optimize(optimizer, item, output, autotune);
}

Status OptimizeWithInjectPrefetchEligible(const GrapplerItem &item,
                                          GraphDef *output, bool autotune) {
  InjectPrefetchEligible optimizer;
  return Optimize(optimizer, item, output, autotune);
}

class InjectPrefetchParameterizedTest : public ::testing::TestWithParam<bool> {
};

TEST_P(InjectPrefetchParameterizedTest, TestAutotuneSetting) {
  const bool autotune = GetParam();

  GrapplerItem item;
  item.graph = test::function::GDef(
      {NDef("start", "Const", {}, {{"value", 0}, {"dtype", DT_INT32}}),
       NDef("stop", "Const", {}, {{"value", 10}, {"dtype", DT_INT32}}),
       NDef("step", "Const", {}, {{"value", 1}, {"dtype", DT_INT32}}),
       NDef("range", "RangeDataset", {"start", "stop", "step"},
            {{"output_shapes", gtl::ArraySlice<TensorShape>{}},
             {"output_types", gtl::ArraySlice<DataType>{}}}),
       NDef("Sink", "Identity", {"range"}, {})});

  item.fetch.push_back("Sink");

  // Test inject_prefetch
  GraphDef inject_prefetch_output;
  TF_ASSERT_OK(
      OptimizeWithInjectPrefetch(item, &inject_prefetch_output, autotune));
  EXPECT_EQ(autotune, graph_utils::ContainsNodeWithOp(kPrefetchDataset,
                                                      inject_prefetch_output));
  EXPECT_EQ(autotune, graph_utils::ContainsGraphNodeWithName(
                          "inject/prefetch_range", inject_prefetch_output));

  // Test inject_prefetch_eligible
  GraphDef inject_prefetch_eligible_output;
  TF_ASSERT_OK(OptimizeWithInjectPrefetchEligible(
      item, &inject_prefetch_eligible_output, autotune));
  EXPECT_EQ(false, graph_utils::ContainsNodeWithOp(
                       kPrefetchDataset, inject_prefetch_eligible_output));
  EXPECT_EQ(false,
            graph_utils::ContainsGraphNodeWithName(
                "inject/prefetch_range", inject_prefetch_eligible_output));
  EXPECT_EQ(item.graph.DebugString(),
            inject_prefetch_eligible_output.DebugString());
}

INSTANTIATE_TEST_SUITE_P(AutotuneSetting, InjectPrefetchParameterizedTest,
                         ::testing::Values(false, true));

TEST(InjectPrefetchTest, FromFunctionDef) {
  GrapplerItem item;
  item.graph = test::function::GDef(
      {NDef("start", "Const", {}, {{"value", 0}, {"dtype", DT_INT32}}),
       NDef("stop", "Const", {}, {{"value", 10}, {"dtype", DT_INT32}}),
       NDef("step", "Const", {}, {{"value", 1}, {"dtype", DT_INT32}}),
       NDef("range", "RangeDataset", {"start", "stop", "step"},
            {{"output_shapes", gtl::ArraySlice<TensorShape>{}},
             {"output_types", gtl::ArraySlice<DataType>{}}}),
       NDef("Sink", "_Retval", {"range"}, {})});

  item.fetch.push_back("Sink");

  GraphDef output;
  TF_ASSERT_OK(OptimizeWithInjectPrefetch(item, &output, true));
  EXPECT_FALSE(graph_utils::ContainsNodeWithOp(kPrefetchDataset, output));
}

TEST(InjectPrefetchTest, AlreadyPrefetched) {
  GrapplerItem item;
  item.graph = test::function::GDef(
      {NDef("start", "Const", {}, {{"value", 0}, {"dtype", DT_INT32}}),
       NDef("stop", "Const", {}, {{"value", 10}, {"dtype", DT_INT32}}),
       NDef("step", "Const", {}, {{"value", 1}, {"dtype", DT_INT32}}),
       NDef("range", "RangeDataset", {"start", "stop", "step"},
            {{"output_shapes", gtl::ArraySlice<TensorShape>{}},
             {"output_types", gtl::ArraySlice<DataType>{}}}),
       NDef("prefetch", kPrefetchDataset, {"range"}, {}),
       NDef("Sink", "Identity", {"prefetch"}, {})});

  item.fetch.push_back("Sink");

  GraphDef output;
  TF_ASSERT_OK(OptimizeWithInjectPrefetch(item, &output, true));
  EXPECT_TRUE(graph_utils::ContainsNodeWithOp(kPrefetchDataset, output));
  EXPECT_EQ(6, output.node_size());
}

TEST(InjectPrefetchTest, AlreadyParallelMap) {
  GrapplerItem item;
  item.graph = test::function::GDef(
      {NDef("start", "Const", {}, {{"value", 0}, {"dtype", DT_INT32}}),
       NDef("stop", "Const", {}, {{"value", 10}, {"dtype", DT_INT32}}),
       NDef("step", "Const", {}, {{"value", 1}, {"dtype", DT_INT32}}),
       NDef("range", "RangeDataset", {"start", "stop", "step"},
            {{"output_shapes", gtl::ArraySlice<TensorShape>{}},
             {"output_types", gtl::ArraySlice<DataType>{}}}),
       NDef("parallel_map", kParallelMapDataset, {"range"},
            {{"f", "__inference_Dataset_map_normalize_8232"},
             {"output_shapes", gtl::ArraySlice<TensorShape>{}},
             {"output_types", gtl::ArraySlice<DataType>{}}}),
       NDef("Sink", "Identity", {"parallel_map"}, {})});

  item.fetch.push_back("Sink");

  GraphDef output;
  TF_ASSERT_OK(OptimizeWithInjectPrefetch(item, &output, true));
  EXPECT_FALSE(graph_utils::ContainsNodeWithOp(kPrefetchDataset, output));
  EXPECT_EQ(6, output.node_size());
}

TEST(InjectPrefetchTest, OptionsFollowedByPrefetched) {
  GrapplerItem item;
  item.graph = test::function::GDef(
      {NDef("start", "Const", {}, {{"value", 0}, {"dtype", DT_INT32}}),
       NDef("stop", "Const", {}, {{"value", 10}, {"dtype", DT_INT32}}),
       NDef("step", "Const", {}, {{"value", 1}, {"dtype", DT_INT32}}),
       NDef("range", "RangeDataset", {"start", "stop", "step"},
            {{"output_shapes", gtl::ArraySlice<TensorShape>{}},
             {"output_types", gtl::ArraySlice<DataType>{}}}),
       NDef("prefetch", kPrefetchDataset, {"range"},
            {{"output_shapes", gtl::ArraySlice<TensorShape>{}},
             {"output_types", gtl::ArraySlice<DataType>{}}}),
       NDef("options", kOptionsDataset, {"prefetch"},
            {{"output_shapes", gtl::ArraySlice<TensorShape>{}},
             {"output_types", gtl::ArraySlice<DataType>{}}}),
       NDef("Sink", "Identity", {"options"}, {})});

  item.fetch.push_back("Sink");

  GraphDef output;
  TF_ASSERT_OK(OptimizeWithInjectPrefetch(item, &output, true));
  EXPECT_FALSE(graph_utils::ContainsGraphNodeWithName("inject/prefetch_options",
                                                      output));
  EXPECT_EQ(7, output.node_size());
}

}  // namespace
}  // namespace grappler
}  // namespace tensorflow
